ash/src/vk.rs

```diff
@@ -82,6 +66,103 @@ pub(crate) unsafe fn ptr_chain_iter<T: ?Sized>(
+    /// valid extension structs can be pushed into the chain.
+    /// If the chain looks like `A -> B -> C` and `D -> E`, and you call `A.extend(&mut D)`,
+    /// then the chain will look like `A -> D -> E -> B -> C`.
+    ///
+    /// # Safety
+    /// This function will walk the [`BaseOutStructure::p_next`] chain of `next`, requiring
+    /// all non-`NULL` pointers to point to a valid Vulkan structure starting with the
+    /// [`BaseOutStructure`] layout.
+    ///
+    /// The last struct in this chain (i.e. the one where `p_next` is `NULL`) must
+    /// be writable memory, as its `p_next` field will be updated with the value of
+    /// `self.p_next`.
+    unsafe fn extend<T: Extends<'a, Self> + ?Sized>(mut self, next: &'a mut T) -> Self
+    where
+        Self: Sized,
+    {
+        // `next` here can contain a pointer chain. This means that we must correctly
+        // attach he head to the root and the tail to the rest of the chain
+        // For example:
+        //
+        // next = A -> B
+        // Before: `Root -> C -> D -> E`
+        // After: `Root -> A -> B -> C -> D -> E`
+        //                 ^^^^^^
+        //                 next chain
+        let slf_base = unsafe { &mut *<*mut _>::cast::<BaseOutStructure<'_>>(&mut self) };
+        let next_base = <*mut T>::cast::<BaseOutStructure<'_>>(next);
+        let last_next = ptr_chain_iter(next).last().unwrap();
+        (*last_next).p_next = slf_base.p_next;
+        slf_base.p_next = next_base;
+        self
+    }
+
+    #[doc(hidden)]
+    #[deprecated = "Migrate to `push()` if `next` does not have an existing chain (i.e. `p_next` is `NULL`), `extend()` otherwise"]
+    unsafe fn push_next<T: Extends<'a, Self> + ?Sized>(self, next: &'a mut T) -> Self
+    where
+        Self: Sized,
+    {
+        self.extend(next)
+    }
+}
+
+/// Non-object-safe variant of [`AnyTaggedStructure`], meaning that a `dyn TaggedStructure` cannot
+/// exist but as a consequence the [`TaggedStructure::STRUCTURE_TYPE`] associated constant is
+/// available.
+///
+/// [`AnyTaggedStructure`]s have a [`BaseInStructure::s_type`] field indicating its type, which must
+/// always match the value of [`TaggedStructure::STRUCTURE_TYPE`].
+pub unsafe trait TaggedStructure<'a>: AnyTaggedStructure<'a> {
+    const STRUCTURE_TYPE: StructureType;
+}
+unsafe impl<'a, T: TaggedStructure<'a>> AnyTaggedStructure<'a> for T {}
+
+/// Implemented for every structure that extends base structure `B`. Concretely that means struct
+/// `B` is listed in its array of [`structextends` in the Vulkan registry][1].
+///
+/// Similar to [`TaggedStructure`], all `unsafe` implementers of this trait must guarantee that
+/// their structure is layout-compatible [`BaseInStructure`] and [`BaseOutStructure`].
+///
+/// [1]: https://registry.khronos.org/vulkan/specs/latest/styleguide.html#extensions-interactions
+pub unsafe trait Extends<'a, B: AnyTaggedStructure<'a>>: AnyTaggedStructure<'a> {}
```

Comment:
Note to self: play with these lifetimes to see if we can simplify any further before we commit.